internal/lifecycle/readme.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lifecycle
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/util"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var (
+	// The tag that should appear immediately before code blocks in a README to indicate that the enclosed commands
+	// are to be used by this program for building and deploying the sample.
+	codeTag = "{sst-run-unix}"
+
+	gcloudCommandRegexp   = regexp.MustCompile(`\bgcloud\b`)
+	cloudRunCommandRegexp = regexp.MustCompile(`\brun\b`)
+
+	gcrURLRegexp = regexp.MustCompile(`gcr.io/.+/\S+`)
+)
+
+// parseREADME parses a README file with the given name. It reads terminal commands surrounded by one of the codeTags
+// listed above and loads them into a Lifecycle. In the process, it replaces the Cloud Run service name and Container
+// Registry tag with the provided inputs.
+func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] os.Open %s: %w", filename, err)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	var lifecycle Lifecycle
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.Contains(line, codeTag) {
+			if s := scanner.Scan(); !s {
+				if err := scanner.Err(); err != nil && !s {
+					return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+				}
+				return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: unexpected EOF; file ended" +
+					"immediately after code tag")
+			}
+
+			startCodeBlockLine := scanner.Text()
+			c := strings.Contains(startCodeBlockLine, "```")
```

Comment:
I'd also use a constant for the triple backticks (and any other magic values like this).

Reply:
Resolved, thank you.